signac/contrib/project.py

```diff
@@ -196,7 +218,12 @@ class Project:
     def __init__(self, config=None, _ignore_schema_version=False):
         if config is None:
             config = load_config()
-        self._config = _ProjectConfig(config)
+        self._config = _ProjectConfig(config, _mutate_hook=partial(_mutate_hook, self))
```

Comment:
I don&rsquo;t understand this line `self._config = _ProjectConfig(config, _mutate_hook=partial(_mutate_hook, self))`. I know that partial will cause the `_mutate_hook` argument of the `_ProjectConfig` to be `_mutate_hook` applied to itself, but where is `_mutate_hook` defined? It seems not defined in this scope. Does it take the value of `_mutate_hook` of the `self` within `_ProjectConfig`?

Reply:
I'll break it down: `_ProjectConfig(config, ...)` calls the constructor of the `_ProjectConfig` class. We provide a keyword argument named `_mutate_hook` with the value `partial(_mutate_hook, self)`. That partial takes the function `_mutate_hook` (now renamed to `_invalidate_config_cache`) and binds the first parameter of that function to `self`. The `partial` returns a callable with no free parameters.

This could equivalently be written as:
```python
func = lambda: _invalidate_config_cache(self)

self._config = _ProjectConfig(config, _mutate_hook=func)
```
or
```python
def func():
    _invalidate_config_cache(self)

self._config = _ProjectConfig(config, _mutate_hook=func)
```
where `self` is defined in the scope of `Project.__init__` and that scope is shared when the function is passed along.

Reply:
thank you!